### PR TITLE
[#T5886] custom_background: added new model and configuration for set the custom background per company per lang

### PR DESCRIPTION
**Pull request by bizzappdevteam**
- Added a new selection in the 'Custom Report Type' at the report level.
- Added new model 'report.company.background.lang' to set the per company and per lang custom background.
- Based on the configuration of the 'Company Background Per Language' the report background is set as per the company and per language.

### CODE REVIEW

File: custom_background/models/report.py [hunk 1]
```diff
@@ -77,6 +77,8 @@ class IrActionsReport(models.Model):
             ("company", "From Company"),
             ("report", "From Report Fixed"),
             ("dynamic", "From Report Dynamic"),
+            # Added new custom report type #T5886
+            ("dynamic_company_per_lang", "Company Background Per Language"),
```
Comment (bizzappdev) [inline custom_background/models/report.py]:
```suggestion
            ("dynamic_per_report_company_lang", "Background Per Report - Company - Lang"),
```

File: custom_background/models/report.py [hunk 2]
```diff
@@ -92,6 +94,12 @@ class IrActionsReport(models.Model):
     is_bg_per_lang = fields.Boolean(
         string="Is Background Per Language",
     )
+    # New field #T5886
+    company_bg_per_lang_ids = fields.One2many(
```
Comment (bizzappdev) [inline custom_background/models/report.py]:
```suggestion
    per_report_com_lang_bg_ids = fields.One2many(
```

File: custom_background/models/report.py [hunk 3]
```diff
@@ -92,6 +94,12 @@ class IrActionsReport(models.Model):
...
+        "report.company.background.lang",
+        "report_id",
+        string="Company Background Per Language",
```
Comment (bizzappdev) [inline custom_background/models/report.py]:
```suggestion
        string="Per Report Company Language Background ",
```

File: custom_background/models/report.py [hunk 4]
```diff
@@ -241,13 +249,42 @@ def get_bg_per_lang(self):
...
         custom_bg_lang = custom_bg_from.bg_per_lang_ids.filtered(
-            lambda l: l.lang_id.code == lang_code
+            lambda lang: lang.lang_id.code == lang_code
         )
 
         # Set 1st custom background.
         custom_background = custom_bg_lang[:1].background_pdf
         return custom_background
 
+    def _get_background_per_company_per_lang(self):
```
Comment (bizzappdev) [inline custom_background/models/report.py]: change name accodringly

File: custom_background/models/report.py [hunk 5]
```diff
@@ -241,13 +249,42 @@ def get_bg_per_lang(self):
...
+        """New method for get the custom background based on the report configuration
+        based on the per company and per Lang. #T5886"""
+        self.ensure_one()
+        lang_code = self.get_lang()
+        company = self._context.get("background_company")
```
Comment (bizzappdev) [inline custom_background/models/report.py:272]: pass company from call

File: custom_background/models/report.py [hunk 6]
```diff
@@ -241,13 +249,42 @@ def get_bg_per_lang(self):
...
+        if custom_bg_only_with_company:
+            return custom_bg_only_with_company[:1].background_pdf
+
+        # Get the Lang matched but company is not set. #T5886
+        custom_bg_only_with_lang = self.company_bg_per_lang_ids.filtered(
+            lambda bg: bg.lang_id.code == lang_code and not bg.company_id
+        )
+        if custom_bg_only_with_lang:
+            return custom_bg_only_with_lang[:1].background_pdf
+        return False
```
Comment (bizzappdev) [inline custom_background/models/report.py:301]:
```suggestion
        default_custom_bg = self.company_bg_per_lang_ids.filtered(
            lambda bg: not bg.lang_id and not bg.company_id
         if default_custom_bg:
             default_custom_bg[:1].background_pdf
        return False
```

File: custom_background/models/report.py [hunk 7]
```diff
@@ -533,6 +570,39 @@ def _run_wkhtmltopdf(  # noqa: C901
                 output.write(open(temp_report_path, "wb"))
                 pdf_report_path = temp_report_path
                 os.close(temp_report_id)
+            # Added new condition for set the background based on the per company and
+            # per Lang. #T5886
+            elif (
+                report
+                and report.custom_report_background
+                and report.is_bg_per_lang
```
Comment (bizzappdev) [inline custom_background/models/report.py]: Do not thing is_bg_per_lang is needed if case of dynamic_company_per_lang is selected

**Review by bizzappdev — CHANGES_REQUESTED** [review]

File: custom_background/models/report.py [hunk 8]
```diff
@@ -603,7 +662,7 @@ def _run_wkhtmltopdf(  # noqa: C901
         for temporary_file in temporary_files:
             try:
                 os.unlink(temporary_file)
-            except (OSError, IOError):  # noqa: disable=B014
```
Comment (bizzappdev) [inline custom_background/models/report.py:606]: why remove?

File: custom_background/models/report_company_background_lang.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    _description = "Report Company Background Line Per Language"
+
+    # New fields. #T5886
+    lang_id = fields.Many2one(
+        "res.lang",
+        string="Language",
+    )
+    background_pdf = fields.Binary(string="Background PDF", required=True)
+    file_name = fields.Char()
+    report_id = fields.Many2one("ir.actions.report", string="Report")
```
Comment (bizzappdev) [inline custom_background/models/report_company_background_lang.py]: ondelete cascade

File: custom_background/models/report_company_background_lang.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    company_id = fields.Many2one(
```
Comment (bizzappdev) [inline custom_background/models/report_company_background_lang.py]: ondelete cascade

File: custom_background/models/report_company_background_lang.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
+# See LICENSE file for full copyright and licensing details.
+from odoo import fields, models
+
+
+class ReportCompanyBackgroundLang(models.Model):
+    _name = "report.company.background.lang"
+    _description = "Report Company Background Line Per Language"
+
+    # New fields. #T5886
+    lang_id = fields.Many2one(
```
Comment (bizzappdev) [inline custom_background/models/report_company_background_lang.py]: ondelete cascade

**Review by bizzappdev — CHANGES_REQUESTED** [review]
Small changes and one question

**Comment by bizzappdev** [comment]
also make pre-commit happy.

File: custom_background/models/report.py (hunk 8 above)
Comment (bizzappdevteam) [reply custom_background/models/report.py:606]: it's removed automatically from the pre-commit

**Review by bizzappdev — APPROVED** [review]

**Review by bizzappdev — APPROVED** [review]

**Pull request merged**